File: src/main/java/org/gradle/api/plugins/shadow/resource/ApacheNoticeResourceTransformer.java
package org.gradle.api.plugins.shadow.resource;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import org.gradle.api.plugins.shadow.relocation.Relocator;
import org.codehaus.plexus.util.StringUtils;

import java.io.*;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.jar.JarEntry;
import java.util.jar.JarOutputStream;

/**
 * Merges <code>META-INF/NOTICE.TXT</code> files.
 */
public class ApacheNoticeResourceTransformer
    implements ResourceTransformer
{
    Set<String> entries = new LinkedHashSet<String>();

    Map<String, Set<String>> organizationEntries = new LinkedHashMap<String, Set<String>>();

    String projectName = ""; // MSHADE-101 :: NullPointerException when projectName is missing

    boolean addHeader = true;

    String preamble1 = "// ------------------------------------------------------------------\n"
        + "// NOTICE file corresponding to the section 4d of The Apache License,\n"
        + "// Version 2.0, in this case for ";

    String preamble2 = "\n// ------------------------------------------------------------------\n";

    String preamble3 = "This product includes software developed at\n";

    //defaults overridable via config in pom
    String organizationName = "The Apache Software Foundation";

    String organizationURL = "http://www.apache.org/";

    String inceptionYear = "2006";

    String copyright;

    /**
     * The file encoding of the <code>NOTICE</code> file.
     */
    String encoding;

    private static final String NOTICE_PATH = "META-INF/NOTICE";

    private static final String NOTICE_TXT_PATH = "META-INF/NOTICE.txt";

    public boolean canTransformResource( String resource )
    {
        if ( NOTICE_PATH.equalsIgnoreCase( resource ) || NOTICE_TXT_PATH.equalsIgnoreCase( resource ) )
        {
            return true;
        }

        return false;
    }

    public void processResource( String resource, InputStream is, List<Relocator> relocators )
        throws IOException
    {
        if ( entries.isEmpty() )
        {
            String year = new SimpleDateFormat( "yyyy" ).format( new Date() );
            if ( !inceptionYear.equals( year ) )
            {
                year = inceptionYear + "-" + year;
            }

            //add headers
            if ( addHeader )
            {
                entries.add( preamble1 + projectName + preamble2 );
            }
            else
            {
                entries.add( "" );
            }
            //fake second entry, we'll look for a real one later
            entries.add( projectName + "\nCopyright " + year + " " + organizationName + "\n" );
            entries.add( preamble3 + organizationName + " (" + organizationURL + ").\n" );
        }

        BufferedReader reader;
        if ( StringUtils.isNotEmpty( encoding ) )
        {
            reader = new BufferedReader( new InputStreamReader( is, encoding ) );
        }
        else
        {
            reader = new BufferedReader( new InputStreamReader( is ) );
        }

        String line = reader.readLine();
        StringBuffer sb = new StringBuffer();
        Set<String> currentOrg = null;
        int lineCount = 0;
        while ( line != null )
        {
            String trimedLine = line.trim();

            if ( !trimedLine.startsWith( "//" ) )
            {
                if ( trimedLine.length() > 0 )
                {
                    if ( trimedLine.startsWith( "- " ) )
                    {
                        //resource-bundle 1.3 mode
                        if ( lineCount == 1
                            && sb.toString().indexOf( "This product includes/uses software(s) developed by" ) != -1 )
                        {
                            currentOrg = organizationEntries.get( sb.toString().trim() );
                            if ( currentOrg == null )
                            {
                                currentOrg = new TreeSet<String>();
                                organizationEntries.put( sb.toString().trim(), currentOrg );
                            }
                            sb = new StringBuffer();
                        }
                        else if ( sb.length() > 0 && currentOrg != null )
                        {
                            currentOrg.add( sb.toString() );
                            sb = new StringBuffer();
                        }

                    }
                    sb.append( line ).append( "\n" );
                    lineCount++;
                }
                else
                {
                    String ent = sb.toString();
                    if ( ent.startsWith( projectName ) && ent.indexOf( "Copyright " ) != -1 )
                    {
                        copyright = ent;
                    }
                    if ( currentOrg == null )
                    {
                        entries.add( ent );
                    }
                    else
                    {
                        currentOrg.add( ent );
                    }
                    sb = new StringBuffer();
                    lineCount = 0;
                    currentOrg = null;
                }
            }

            line = reader.readLine();
        }
        if ( sb.length() > 0 )
        {
            if ( currentOrg == null )
            {
                entries.add( sb.toString() );
            }
            else
            {
                currentOrg.add( sb.toString() );
            }
        }
    }

    public boolean hasTransformedResource()
    {
        return true;
    }

    public void modifyOutputStream( JarOutputStream jos )
        throws IOException
    {
        jos.putNextEntry( new JarEntry( NOTICE_PATH ) );

        Writer pow;
        if ( StringUtils.isNotEmpty( encoding ) )
        {
            pow = new OutputStreamWriter( jos, encoding );
        }
        else
        {
            pow = new OutputStreamWriter( jos );
        }
        PrintWriter writer = new PrintWriter( pow );

        int count = 0;
        for ( String line : entries )
        {
            ++count;
            if ( line.equals( copyright ) && count != 2 )
            {
                continue;
            }

            if ( count == 2 && copyright != null )
            {
                writer.print( copyright );
                writer.print( '\n' );
            }
            else
            {
                writer.print( line );
                writer.print( '\n' );
            }
            if ( count == 3 )
            {
                //do org stuff
                for ( Map.Entry<String, Set<String>> entry : organizationEntries.entrySet() )
                {
                    writer.print( entry.getKey() );
                    writer.print( '\n' );
                    for ( String l : entry.getValue() )
                    {
                        writer.print( l );
                    }
                    writer.print( '\n' );
                }
            }
        }

        writer.flush();

        entries.clear();
    }
}

File: src/main/kotlin/com/github/jengelman/gradle/plugins/shadow/transformers/ApacheNoticeResourceTransformer.kt
package com.github.jengelman.gradle.plugins.shadow.transformers

import com.github.jengelman.gradle.plugins.shadow.internal.property
import java.io.PrintWriter
import java.nio.charset.Charset
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TreeSet
import javax.inject.Inject
import org.apache.tools.zip.ZipEntry
import org.apache.tools.zip.ZipOutputStream
import org.gradle.api.file.FileTreeElement
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.Property
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.Optional

/**
 * Merges `META-INF/NOTICE.TXT` files.
 *
 * Modified from [org.apache.maven.plugins.shade.resource.ApacheNoticeResourceTransformer.java](https://github.com/apache/maven-shade-plugin/blob/master/src/main/java/org/apache/maven/plugins/shade/resource/ApacheNoticeResourceTransformer.java).
 *
 * @author John Engelman
 */
public open class ApacheNoticeResourceTransformer @Inject constructor(
  final override val objectFactory: ObjectFactory,
) : Transformer {
  private val entries = mutableSetOf<String>()
  private val organizationEntries = mutableMapOf<String, MutableSet<String>>()
  private inline val charset get() = Charset.forName(charsetName.get())

  @get:Input
  public open val projectName: Property<String> = objectFactory.property("")

  @get:Input
  public open val addHeader: Property<Boolean> = objectFactory.property(true)

  @get:Input
  public open val preamble1: Property<String> = objectFactory.property(
    """
      // ------------------------------------------------------------------
      // NOTICE file corresponding to the section 4d of The Apache License,
      // Version 2.0, in this case for
    """.trimIndent(),
  )

  @get:Input
  public open val preamble2: Property<String> = objectFactory.property(
    "\n// ------------------------------------------------------------------\n",
  )

  @get:Input
  public open val preamble3: Property<String> = objectFactory.property("This product includes software developed at\n")

  @get:Input
  public open val organizationName: Property<String> = objectFactory.property("The Apache Software Foundation")

  @get:Input
  public open val organizationURL: Property<String> = objectFactory.property("http://www.apache.org/")

  @get:Input
  public open val inceptionYear: Property<String> = objectFactory.property("2006")

  @get:Optional
  @get:Input
  public open val copyright: Property<String> = objectFactory.property()

  /**
   * The file encoding of the `NOTICE` file.
   */
  @get:Input
  public open val charsetName: Property<String> = objectFactory.property(Charsets.UTF_8.name())

  override fun canTransformResource(element: FileTreeElement): Boolean {
    val path = element.relativePath.pathString
    return NOTICE_PATH.equals(path, ignoreCase = true) ||
      NOTICE_TXT_PATH.equals(path, ignoreCase = true) ||
      NOTICE_MD_PATH.equals(path, ignoreCase = true)
  }

  override fun transform(context: TransformerContext) {
    val projectName = projectName.get()
    val addHeader = addHeader.get()
    val preamble1 = preamble1.get()
    val preamble2 = preamble2.get()
    val preamble3 = preamble3.get()
    val organizationName = organizationName.get()
    val organizationURL = organizationURL.get()
    val inceptionYear = inceptionYear.get()

    if (entries.isEmpty()) {
      val year = SimpleDateFormat("yyyy", Locale.US).format(Date()).let {
        if (inceptionYear != it) "$inceptionYear-$it" else it
      }
      // add headers
      if (addHeader) {
        entries.add("$preamble1$projectName$preamble2")
      } else {
        entries.add("")
      }
      // fake second entry, we'll look for a real one later
      entries.add("$projectName\nCopyright $year $organizationName\n")
      entries.add("$preamble3$organizationName ($organizationURL).\n")
    }

    val reader = context.inputStream.bufferedReader(charset)
    var line = reader.readLine()
    val sb = StringBuffer()
    var currentOrg: MutableSet<String>? = null
    var lineCount = 0
    while (line != null) {
      val trimmedLine = line.trim()
      if (!trimmedLine.startsWith("//")) {
        if (trimmedLine.isNotEmpty()) {
          if (trimmedLine.startsWith("- ")) {
            // resource-bundle 1.3 mode
            if (lineCount == 1 && sb.toString().contains("This product includes/uses software(s) developed by")) {
              currentOrg = organizationEntries.getOrPut(sb.toString().trim()) { TreeSet() }
              sb.setLength(0)
            } else if (sb.isNotEmpty() && currentOrg != null) {
              currentOrg.add(sb.toString())
              sb.setLength(0)
            }
          }
          sb.append(line).append("\n")
          lineCount++
        } else {
          val entry = sb.toString()
          if (entry.startsWith(projectName) && entry.contains("Copyright ")) {
            copyright.set(entry)
          }
          if (currentOrg == null) {
            entries.add(entry)
          } else {
            currentOrg.add(entry)
          }
          sb.setLength(0)
          lineCount = 0
          currentOrg = null
        }
      }

      line = reader.readLine()
    }
    if (sb.isNotEmpty()) {
      if (currentOrg == null) {
        entries.add(sb.toString())
      } else {
        currentOrg.add(sb.toString())
      }
    }
  }

  override fun hasTransformedResource(): Boolean = true

  override fun modifyOutputStream(os: ZipOutputStream, preserveFileTimestamps: Boolean) {
    val copyright = copyright.orNull

    val zipEntry = ZipEntry(NOTICE_PATH)
    zipEntry.time = TransformerContext.getEntryTimestamp(preserveFileTimestamps, zipEntry.time)
    os.putNextEntry(zipEntry)

    val writer = PrintWriter(os.writer(charset))

    var count = 0
    for (line in entries) {
      count++
      if (line == copyright && count != 2) continue
      if (count == 2 && copyright != null) {
        writer.print(copyright)
        writer.print('\n')
      } else {
        writer.print(line)
        writer.print('\n')
      }
      if (count == 3) {
        // do org stuff
        for ((key, value) in organizationEntries) {
          writer.print(key)
          writer.print('\n')
          for (l in value) {
            writer.print(l)
          }
          writer.print('\n')
        }
      }
    }

    writer.flush()
    entries.clear()
  }

  private companion object {
    private const val NOTICE_PATH = "META-INF/NOTICE"
    private const val NOTICE_TXT_PATH = "META-INF/NOTICE.txt"
    private const val NOTICE_MD_PATH = "META-INF/NOTICE.md"
  }
}
